File: AtCoder/ABC023/d.cs
using System;
using System.Linq;
using System.Collections;
using System.Collections.Generic;

namespace ABC023
{
    class Balloon: IComparable
    {
        public long H;
        public long S;
        public long D;
        public Balloon(long H, long S, long D)
        {
            this.H = H;
            this.S = S;
            this.D = D;
        }

        public int CompareTo(object obj)
        {
            Balloon i = obj as Balloon;
            return this.D.CompareTo(i.D);
        }
    }
    class Program
    {
        static bool Check(Balloon[] B, long val)
        {
            long[] h = new long[B.Length];
            for (int i = 0; i < B.Length; i++)
            {
                if (val < B[i].H) return false;
                h[i] = (val - B[i].H) / B[i].S;
            }
            Array.Sort(h);
            for (int i = 0; i < B.Length; i++)
            {
                if (h[i] < i) return false;
            }
            return true;
        }

        static long BinarySearch(Balloon[] B, long left, long right)
        {
            while (left < right)
            {
                long mid = (left + right) / 2;
                bool f = Check(B, mid);
                if (f)
                {
                    right = mid;
                }
                else
                {
                    left = mid + 1;
                }
            }
            return left;
        }

        public static void Main()
        {
            int N = int.Parse(Console.ReadLine());
            Balloon[] B = new Balloon[N];
            long m = 0;
            for (int i = 0; i < N; i++)
            {
                string[] s = Console.ReadLine().Split(' ');
                long H = long.Parse(s[0]);
                long S = long.Parse(s[1]);
                long D = H + (N - 1) * S;
                m = Math.Max(m, D);
                B[i] = new Balloon(H, S, D);
            }
            // Array.Sort(B);
            long res = BinarySearch(B, 1, m + 1);
            Console.WriteLine(res);
        }
    }
}

File: AtCoder/ABC023/d.cpp
#include <iostream>
#include <algorithm>

using namespace std;

bool check(long* h, long* s, long x, int N)
{
    long t[N];
    for (int i = 0; i < N; i++)
    {
        if (x < h[i]) return false;
        t[i] = (x - h[i]) / s[i];
    }
    sort(t, t + N);
    for (int i = 0; i < N; i++)
    {
        if (t[i] < i) return false;
    }
    return true;
}

long binary_search(long* h, long* s, long left, long right, int N)
{
    while (left < right)
    {
        long mid = (left + right) / 2;
        if (check(h, s, mid, N))
        {
            right = mid;
        }
        else
        {
            left = mid + 1;
        }
    }
    return left;
}

int main()
{
    int N;
    cin >> N;
    long h[N];
    long s[N];
    long m = 0;
    for (int i = 0; i < N; i++)
    {
        cin >> h[i] >> s[i];
        long d = h[i] + s[i] * (long)(N - 1);
        if (d >= m)
        {
            m = d;
        }
    }
    cout << binary_search(h, s, 1, m + 1, N) << "\n";
    return 0;
}